### DISCUSSION TITLE
Install binaries when possible

### DISCUSSION THREAD

**Discussion by cig0** [Homebrew/discussions#2790] [Linux] [answered]
Hi,

Is there a way to tell Homebrew to go ahead and install the binary release for a given formula, when one is available?

The last time I checked my `.linuxbrew` directory it was weighting about 11Gb for some 282 formulas, from which a half or less are apps that offer a precompiled binary. :scream:

If there's a binary executable file available for the app I want to install, I'd like to install it instead of building from the formula, being building using a formula a nice fallback feature in case there's an issue with the provided binary.

Thanks!

**Comment by SMillerDev** [comment]
If you mean Homebrew bottles, Homebrew will try and install those by default.
If you mean that upstream provides a binary, that's not how homebrew works. Everything in Homebrew is compiled from source in CI.

**Comment by carlocab** [comment]
If you install Homebrew into `/home/linuxbrew/.linuxbrew` then over 80% of formulae you install should be precompiled.

**Comment by gromgit** [comment]
> The last time I checked my `.linuxbrew` directory it was weighting about 11Gb for some 282 formulas, from which a half or less are apps that offer a precompiled binary. :scream:

Since you seem to be looking at the _total_ size of `/home/linuxbrew/.linuxbrew`, that's not unusual. The `var` subdirectory, in particular, contains the data directories for PostgreSQL, MySQL, and pretty much every other database-related formula. In my own setup, `var/cache/fontconfig` is HUGE, weighing in at over 8GB of my total 18GB `.linuxbrew` usage.

Also, if you're building from source a lot, the `gcc` and other tools required to do such builds would definitely take up quite a bit of room. Might be a good idea to investigate why you have to do that so often. What's the output of `brew config` and `brew doctor`?

**Comment by cig0** [comment, marked as answer]
Hi guys,

Thanks for your answers.

So, if I correctly understand what you say, precompiled binaries are only installed if the formula defines a bottle, otherwise, it will build from the source code?

Let´s say I want to install the Terraform client, which is also provided as a pre-compiled executable (by its developers):

```shell
> brew info terraform   
terraform: stable 1.1.3 (bottled), HEAD
Tool to build, change, and version infrastructure
https://www.terraform.io/
Conflicts with:
  tfenv (because tfenv symlinks terraform binaries)
/home/linuxbrew/.linuxbrew/Cellar/terraform/1.1.3 (6 files, 59.1MB) *
  Poured from bottle on 2022-01-14 at 14:28:27
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/terraform.rb
License: MPL-2.0
==> Dependencies
Build: go ✔
Required: gcc ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 45,108 (30 days), 125,417 (90 days), 512,702 (365 days)
install-on-request: 42,318 (30 days), 117,366 (90 days), 482,088 (365 days)
build-error: 7 (30 days)
```

I see that this formula requires both Go and the GCC compiler. If I go ahead and remove those formulas:

```shell
> brew info terraform                    
terraform: stable 1.1.3 (bottled), HEAD
Tool to build, change, and version infrastructure
https://www.terraform.io/
Conflicts with:
  tfenv (because tfenv symlinks terraform binaries)
/home/linuxbrew/.linuxbrew/Cellar/terraform/1.1.3 (6 files, 59.1MB) *
  Poured from bottle on 2022-01-14 at 14:28:27
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/terraform.rb
License: MPL-2.0
==> Dependencies
Build: go ✘
Required: gcc ✘
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 45,108 (30 days), 125,417 (90 days), 512,702 (365 days)
install-on-request: 42,318 (30 days), 117,366 (90 days), 482,088 (365 days)
build-error: 7 (30 days)
```

And then attempt to install Terraform:

```shell
> brew install terraform                        
Running `brew update --preinstall`...
==> Downloading https://ghcr.io/v2/homebrew/core/gcc/manifests/11.2.0_3-1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/gcc/blobs/sha256:3717134ab0f56e7eeb167c4f4a993c81329d6c1248dae5ee6e39f59cfdfa0eee
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:3717134ab0f56e7eeb167c4f4a993c81329d6c1248dae5ee6e39f59cfdfa0eee?se=2022-01-14T17%3A40%3A00Z&sig=rb4Q8%2
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/terraform/manifests/1.1.3
Already downloaded: /home/cig0/.cache/Homebrew/downloads/872edae640e0de9a7c0a524245f8604144de3b29770ac9128b32a3b04ad82df1--terraform-1.1.3.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/terraform/blobs/sha256:6dc578410c3c229cabb61003db1f5e0f0473f74461f482d8498c52597af9925c
Already downloaded: /home/cig0/.cache/Homebrew/downloads/c6a8c53fb6a64b4317ec628b1a2c2a8ee9a63b11626c0cd41c9891abd7e9531e--terraform--1.1.3.x86_64_linux.bottle.tar.gz
==> Installing dependencies for terraform: gcc
==> Installing terraform dependency: gcc
==> Pouring gcc--11.2.0_3.x86_64_linux.bottle.1.tar.gz
==> Creating the GCC specs file: /home/linuxbrew/.linuxbrew/Cellar/gcc/11.2.0_3/bin/../lib/gcc/11/gcc/x86_64-pc-linux-gnu/11/specs
🍺  /home/linuxbrew/.linuxbrew/Cellar/gcc/11.2.0_3: 2,194 files, 348.2MB
==> Installing terraform
==> Pouring terraform--1.1.3.x86_64_linux.bottle.tar.gz
🍺  /home/linuxbrew/.linuxbrew/Cellar/terraform/1.1.3: 6 files, 59.1MB
==> Running `brew cleanup terraform`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
```

I guess that in this case the `terraform` formula doesn't have any bottles defined, so it will compile from the source code.

**Question**:

Should there be any bottles available for this formula, would they be installed then instead of compiling the source code, without me requiring to install its build dependencies?

Thanks,
-Martin

**Reply by gromgit** [reply]
You're reading it wrong.
>```
>> brew info terraform                    
>terraform: stable 1.1.3 (bottled), HEAD
>[...]
>/home/linuxbrew/.linuxbrew/Cellar/terraform/1.1.3 (6 files, 59.1MB) *
>  Poured from bottle on 2022-01-14 at 14:28:27
>```
This says there _is_ a bottle for `terraform`, _and_ it was installed from bottle.
>```
>> brew install terraform                        
>[...]
>==> Pouring gcc--11.2.0_3.x86_64_linux.bottle.1.tar.gz
>[...]
>==> Pouring terraform--1.1.3.x86_64_linux.bottle.tar.gz
>```
This says both `gcc` and `terraform` were installed from bottles.

> Should there be any bottles available for this formula, would they be installed then instead of compiling the source code, without me requiring to install its build dependencies?

Yes. If it goes ahead and builds from source, `brew` will tell you _why_ it decided to do so.

**Reply by carlocab** [reply]
Yea, if you had to compile GCC, you'd notice. It can take up to several hours.

**Reply by cig0** [reply]
Perfect, thanks!
